src/DataMapper/Product/PopularityDataMapper.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusMeilisearchPlugin\DataMapper\Product;
+
+use Doctrine\Persistence\ManagerRegistry;
+use Setono\Doctrine\ORMTrait;
+use Setono\SyliusMeilisearchPlugin\DataMapper\DataMapperInterface;
+use Setono\SyliusMeilisearchPlugin\Document\Document;
+use Setono\SyliusMeilisearchPlugin\Document\Product;
+use Setono\SyliusMeilisearchPlugin\Model\IndexableInterface;
+use Setono\SyliusMeilisearchPlugin\Provider\IndexScope\IndexScope;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Product\Model\ProductVariantInterface;
+use Webmozart\Assert\Assert;
+
+final class PopularityDataMapper implements DataMapperInterface
+{
+    use ORMTrait;
+
+    public function __construct(
+        ManagerRegistry $managerRegistry,
+        /** @var class-string $orderClass */
+        private readonly string $orderClass,
+        /** @var class-string $orderItemClass */
+        private readonly string $orderItemClass,
+        private readonly string $popularityLookBackPeriod = '3 months',
+    ) {
+        $this->managerRegistry = $managerRegistry;
+    }
+
+    /**
+     * @param Product|Document $target
+     * @param array<string, mixed> $context
+     */
+    public function map(IndexableInterface $source, Document $target, IndexScope $indexScope, array $context = []): void
+    {
+        Assert::true(
+            $this->supports($source, $target, $indexScope, $context),
+            'The given $source and $target is not supported',
+        );
+
+        $variants = $source->getEnabledVariants()->map(static fn (ProductVariantInterface $variant): int => (int) $variant->getId())->toArray();
+        if ([] === $variants) {
+            return;
+        }
+
+        $orderIdLowerBound = $this->getOrderIdLowerBound();
+
+        // todo do we need some filters on the order? E.g. state...
```

Comment:
I suppose we should filter only for completed orders, or, even better - for paid orders. If something is returned or not paid, I don't believe we should count as it "popular" 😅

Reply:
It is quite common to see the 'add to cart' event as a big indicator in scenarios like these, however, I agree that returned orders should be a indicator of the opposite, but we don't handle that out of the box in Sylius and hence I can't mirror that in the popularity score.

The reason I didn't do it in the first place is because I need a join to do it and that will likely increase the SQL query time :D